src/socket.io/categories.js

```diff
@@ -171,23 +171,50 @@ SocketCategories.getMoveCategories = function(socket, data, callback) {
 };
 
 SocketCategories.watch = function(socket, cid, callback) {
-	user.watchCategory(socket.uid, cid, function(err) {
-		if (err) {
-			return callback(err);
-		}
-		topics.pushUnreadCount(socket.uid, callback);
-	});
+	ignoreOrWatch(user.watchCategory, socket, cid, callback);
 };
 
 SocketCategories.ignore = function(socket, cid, callback) {
-	user.ignoreCategory(socket.uid, cid, function(err) {
-		if (err) {
-			return callback(err);
-		}
-		topics.pushUnreadCount(socket.uid, callback);
-	});
+	ignoreOrWatch(user.ignoreCategory, socket, cid, callback);
 };
 
+function ignoreOrWatch(fn, socket, cid, callback) {
+	async.waterfall([
+		function(next) {
+			db.getSortedSetRange('categories:cid', 0, -1, next);
+		},
+		function(cids, next) {
+			categories.getCategoriesFields(cids, ['cid', 'parentCid'], next);
+		},
+		function(categoryData, next) {
+			categoryData.forEach(function(c) {
+				c.cid = parseInt(c.cid, 10);
+				c.parentCid = parseInt(c.parentCid, 10);
+			});
+
+			var cids = [parseInt(cid, 10)];
+
+			// filter to subcategories of cid
+
+			var any = true;
+			while (any) {
+				any = false;
+				categoryData.forEach(function(c) {
+					if (cids.indexOf(c.cid) === -1 && cids.indexOf(c.parentCid) !== -1) {
+						cids.push(c.cid);
+						any = true;
+					}
+				});
+			}
+
+			async.each(cids, async.apply(fn, socket.uid), next);
```

Comment:
Can you change this to 

```
async.each(cids, function(cid, next) {
    fn(socket.uid, cid, next);
}, next);
```

I think it's less confusing seeing fn being called with the parameters compared to the async.apply magic.